sdk/core/azure-core/inc/azure/core/io/body_stream.hpp

```diff
@@ -81,6 +81,9 @@ namespace Azure { namespace Core { namespace IO {
         int64_t count,
         Azure::Core::Context const& context = Azure::Core::Context())
     {
+      AZURE_ASSERT_MSG(
+          buffer && count >= 0, "Count cannot be negative, and the buffer pointer cannot be null.");
```

Comment:
@vhvb1989, @CaseyCarter I wanted to construct a bit more rich error message for the customer, but only when the precondition fails.

However, I can't write something like the following, which was possible with exceptions:
```C++
AZURE_ASSERT_MSG(
          buffer && count >= 0, "Count cannot be negative: " + std::to_string(count)
              + ", and the buffer pointer cannot be null.");
```

What do you recommend? Should I do something like:
if (validation)
  create error message
  assert (false, message)

Reply:
I recommend not using a message here.  The message should be used only when we are sure the ASSERT is false. For example for NOT_IMPLEMENTED or for UNREACHABLE_CODE.  In those cases the ASSERT is hardcoded to false and the msg helps to understand the reason.

But for cases like this, we should only call AZURE_ASSERT(exp). The debug message will be clear because it will tell the developer what was evaluated to be true and failed.  Like, in this case:

Debug would be something like "ASSERTION FAILED:  buffer && count >= 0   ....  line xxxx"   It is clear that somehow I am violating the contract and I can see the contract.

For the case of the HARDCODED false, it would be something like "ASSERTION FAILED:  false ,...  line xxx" .  Which is not so good because the contract is not clear.  That's where the msg comes to the rescue :)

Reply:
+1 "don't bother". Wasting more code size for prettier error messages that a properly written program will never see doesn't seem like a good investment. If a user is concerned about the precise failure state they'll reproduce the failure in a debugger and examine whatever state they like regardless of what you put in the error message.